grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/PathQueryImpl.java

```diff
@@ -68,48 +69,119 @@
             throw GraqlQueryException.instanceDoesNotExist();
         }
         if (sourceId.equals(destinationId)) {
-            return Optional.of(Collections.singletonList(tx.get().getConcept(sourceId)));
+            return Collections.singletonList(Collections.singletonList(tx.get().getConcept(sourceId)));
         }
-        ComputerResult result;
 
+        ComputerResult result;
         Set<LabelId> subLabelIds = convertLabelsToIds(subLabels);
-
         try {
             result = getGraphComputer().compute(
-                    new ShortestPathVertexProgram(sourceId, destinationId),
-                    new ClusterMemberMapReduce(ShortestPathVertexProgram.FOUND_IN_ITERATION),
-                    subLabelIds);
+                    new ShortestPathVertexProgram(sourceId, destinationId), null, subLabelIds);
         } catch (NoResultException e) {
             LOGGER.info("ShortestPathVertexProgram is done in " + (System.currentTimeMillis() - startTime) + " ms");
-            return Optional.empty();
+            return Collections.emptyList();
+        }
+
+        Map<String, Set<String>> predecessorMapFromSource = getPredecessorMap(result);
+        List<List<Concept>> allPaths = getAllPaths(predecessorMapFromSource);
+        if (includeAttribute) { // this can be slow
+            return getExtendedPaths(allPaths);
+        }
+
+        LOGGER.info("Number of paths: " + allPaths.size());
+        LOGGER.info("ShortestPathVertexProgram is done in " + (System.currentTimeMillis() - startTime) + " ms");
+        return allPaths;
+    }
+
+    private List<List<Concept>> getExtendedPaths(List<List<Concept>> allPaths) {
```

Comment:
I'd put a Javadoc explaining what this method is for. It's the one for handling attribute edges right?